### PR TITLE
Prefer not to use metaprogramming to DRY up code

### PR DESCRIPTION
**Pull request by zeisler**

### CODE REVIEW

**Comment by mleglise-renew** [comment]
I don't know that we should discourage meta programming this hard. I'm thinking more that if you're going to code it the first way, then you should have a bunch of comments that describe what you're doing, and that list out the full method names that you're generating.

**Comment by zeisler** [comment]
@mleglise-renew I think by the time you go about writing `def method_name` in a comment you might as well just wrote it the normal way. What I am trying to make the point about it's not worth it to meta program just to save a few key strokes. Programming in the manner breaks code navigation in programs like RubyMine and Sublime.

**Comment by zeisler** [comment]
Here is an example from our code base.

https://github.com/projectdx/nemo/blob/master/lib/services/external_request.rb#L20

Finding where these methods are defined is difficult. I would recommend the following:

``` ruby
def get(*args, &block)
  run_request(__method__, *args, &block)
end

def head(*args, &block)
  run_request(__method__, *args, &block)
end

def delete(*args, &block)
  run_request(__method__, *args, &block)
end

def post(*args, &block)
  run_request(__method__, *args, &block)
end

def put(*args, &block)
  run_request(__method__, *args, &block)
end

def patch(*args, &block)
  run_request(__method__, *args, &block)
end
```

The upfront work pays off later. Notice how I use all the non-metaprongramming ruby tools decrease error prone repetition.

**Comment by zeisler** [comment]
To make it clear I really love Ruby's metaprogramming abilities, I use them all the time to do things that I could not do otherwise. 

Here is one example: https://github.com/projectdx/lenny/blob/master/lib/decision_engine/multi_report_source/factory_builder.rb#L8

File: README.md [hunk 1]
```diff
@@ -3610,6 +3610,39 @@ Prefer double-quotes unless your string literal contains `"` or escape character
   Avoid needless metaprogramming.
 <sup>[[link](#no-needless-metaprogramming)]</sup>
 
+  When possible prefer non metaprogramming abstractions to DRY up code because it's more readable and searchable.
+  ```ruby
+  #bad
+  
+  [:account_number, :current_balance, :high_credit,: past_due].each do |method|
```
Comment (hungryed) [inline README.md]: remove space in front of past_due

**Comment by hungryed** [comment]
I dont have a strong preference either way. the ability to use ruby's metaprogramming to define methods is fun and makes adding new ones easy but only for people who already understand the mechanics of how it works. i know i personally skim classes all the time and love to find new tricks i didn't know about but thats not a reason to increase initial overhead

**Comment by natemontgomery** [comment]
I think this kind of thing is very situational.  Either of the solutions (metaprogrammed or explicitly defined) in this case would lead to an additional abstraction of the underlying HTTP client API, which could be useful but we should leverage that underlying API as much as possible.

In the example dustin gave I am only seeing the addition of logging logic around the Faraday request, why would a user not want to use the run_request method directly (making it public of course)?

**Comment by chriswright47** [comment]
I am in favor of preferring non-meta programming as a general rule to improve readability and navigability. I will not block meta-programming commits if they seem reasonable, however :)

**Comment by oscaralanpierce** [comment]
@chriswright47 That's about where I am too.

**Comment by ksharma11** [comment]
👍

**Comment by zeisler** [comment]
@chriswright47 I agree with what you are saying. So is the alteration to the style guide as is a good change or how would you alter it to best fit what your are saying. 

My motivation for bring this up is because I want the team to have a shared vision on this to ease our collaboration.

**Comment by davidbegin** [comment]
👍 I am ok for this as the default. As long as we continue to allow exceptions based on code review conversations!

**Pull request merged**